### PR TITLE
[serve] Initial version of `local_testing_mode`

### PR DESCRIPTION
**Pull request by edoakes**
## Why are these changes needed?

Adds initial support for a (currently private) `_local_testing_mode` flag to `serve.run`. The intention for this flag is to enable users to write unit tests for their application and model composition logic.

User code for each deployment will be run in a background thread using the `UserCallableWrapper` that runs in replica actors. There is a new `Router` and `ReplicaResult` implementation that interacts with the user code to enable the existing `DeploymentHandle` code to work.

Before merging:

- [x] Figure out why `RAY_SERVE_FORCE_LOCAL_TESTING_MODE=1 pytest -vs tests/test_handle_cancellation.py` hangs locally

Follow-ups to this PR:

- Fix blocking .result() in resolving args for composition
- Make get_replica_context() and other auxiliary APIs work
- Support FastAPI TestClient

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/serve/tests/unit/test_local_testing_mode.py [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+
+import pytest
+
+from ray import serve
+from ray.serve.handle import DeploymentHandle
+
+# TODO(edoakes):
+# - add test for to_object_ref error.
+# - support local http client.
+# - test handle.options things.
```
Comment (GeneDer) [inline python/ray/serve/tests/unit/test_local_testing_mode.py:11]: Maybe we can try those in GH issues 🙃

**Review by GeneDer — APPROVED** [review]
Knowing there are still follow up items, can probably also follow up with doc changes. This PR LGTM 👍

File: python/ray/serve/tests/unit/test_local_testing_mode.py (hunk 1 above)
Comment (edoakes) [reply python/ray/serve/tests/unit/test_local_testing_mode.py:11]: yep will be

File: python/ray/serve/_private/test_utils.py [hunk 2]
```diff
@@ -514,14 +515,22 @@ def ping_fruit_stand(channel, app_name):
...
     try:
-        await asyncio.sleep(100000)
+        yield
+        await asyncio.sleep(100)
     except asyncio.CancelledError:
         # Clear the context var to avoid Ray recursively cancelling this method call.
         ray._raylet.async_task_id.set(None)
         await signal_actor.send.remote()
 
+    if not cancelled:
```
[the remark this answers is not in the document]
Comment (edoakes) [reply python/ray/serve/_private/test_utils.py:529]: good catch

File: python/ray/serve/_private/constants.py [hunk 3]
```diff
@@ -350,3 +350,9 @@
 RAY_SERVE_USE_COMPACT_SCHEDULING_STRATEGY = (
     os.environ.get("RAY_SERVE_USE_COMPACT_SCHEDULING_STRATEGY", "0") == "1"
 )
+
+# Always set local_testing_mode=True in serve.run.
```
[the remark this answers is not in the document]
Comment (edoakes) [reply python/ray/serve/_private/constants.py]: updated wording

**Pull request merged**